### ISSUE TITLE
PubSub not reconnecting to all channels

### ISSUE THREAD

**Issue by mail278** [StackExchange/StackExchange.Redis#517]
Hi!

There are few threads with similar issue, but this one is easily reproducible. 
TL;DR
There is a race condition when resubscribing to channels so when you reconnect and then again before everything is reconnected you resubscribe only to part of all channels.

Steps to reproduce
1. Make a lot of subscriptions (1000) to remote server 
- this is important to take long as this has to take more than time in 2.
2. Force connection to reconnect twice in short time (0.5 second)
- I do this by sending a lot of data and filling output buffer on the server (send 100Mbit/s and subscribe 12x - your connection will drop in a short time)

Now if I have channels 1..1000 than I end up not subscribed to some channels from the start (1..N).

I think that bug is in the method of resubscribing after reconnect that this method.

**Comment by NickCraver** [comment]
To clarify: are you saying the last `n` channels are the ones that are resubscribed? e.g. it iterated 400 on the initial reconnect attempt, and switched to the second connection and only the last 600 finished?

If so, that sounds entirely plausible given the locking semantics around it and it being unaware of the connection change in the middle. We'll take a look when in this code for async locking changes in 2.x with the .NET 4.0 drop.

**Comment by NickCraver** [comment]
A few issues overlapped here, the root of the re-sub issue is resolved in 1.2.6. Some ordering issues (correct but confusing) remain - we'll follow up on those. Closing out to cleanup :)
